File: applications/pctfdktwodweights/pctfdktwodweights.cxx
#include "pctfdktwodweights_ggo.h"
#include "rtkGgoFunctions.h"

#include <rtkThreeDCircularProjectionGeometryXMLFile.h>
#include <rtkProjectionsReader.h>
#include "pctFDKDDWeightProjectionFilter.h"

#include <itkImageFileWriter.h>
#include <itkRegularExpressionSeriesFileNames.h>

int main(int argc, char * argv[])
{
  GGO(pctfdktwodweights, args_info);

  typedef float OutputPixelType;
  const unsigned int Dimension = 3;

  // Generate file names
  itk::RegularExpressionSeriesFileNames::Pointer names = itk::RegularExpressionSeriesFileNames::New();
  names->SetDirectory(args_info.path_arg);
  names->SetNumericSort(false);
  names->SetRegularExpression(args_info.regexp_arg);
  names->SetSubMatch(0);

  if(args_info.verbose_flag)
    std::cout << "Regular expression matches "
              << names->GetFileNames().size()
              << " file(s)..."
              << std::endl;

  // Projections reader
  typedef itk::Image< OutputPixelType, Dimension+1 > ProjectionImageType;
  typedef rtk::ProjectionsReader< ProjectionImageType > ReaderType;
  ReaderType::Pointer reader = ReaderType::New();
  reader->SetFileNames( names->GetFileNames() );
  TRY_AND_EXIT_ON_ITK_EXCEPTION( reader->GenerateOutputInformation() );

  // Geometry
  if(args_info.verbose_flag)
    std::cout << "Reading geometry information from "
              << args_info.geometry_arg
              << "..."
              << std::endl;
  rtk::ThreeDCircularProjectionGeometryXMLFileReader::Pointer geometryReader;
  geometryReader = rtk::ThreeDCircularProjectionGeometryXMLFileReader::New();
  geometryReader->SetFilename(args_info.geometry_arg);
  TRY_AND_EXIT_ON_ITK_EXCEPTION( geometryReader->GenerateOutputInformation() )

  // Weights filter
  typedef pct::FDKDDWeightProjectionFilter< ProjectionImageType > WeightType;
  WeightType::Pointer wf = WeightType::New();
  wf->SetInput( reader->GetOutput() );
  wf->SetGeometry( geometryReader->GetOutputObject() );
  wf->InPlaceOff();

  // Write
  typedef itk::ImageFileWriter<  ProjectionImageType > WriterType;
  WriterType::Pointer writer = WriterType::New();
  writer->SetFileName( args_info.output_arg );
  writer->SetInput( wf->GetOutput() );
  writer->SetNumberOfStreamDivisions( args_info.divisions_arg );
  TRY_AND_EXIT_ON_ITK_EXCEPTION( writer->Update() )

  return EXIT_SUCCESS;
}

File: applications/pctfdktwodweights/pctfdktwodweights.py
import argparse

from itk import PCT as pct
from itk import RTK as rtk
import itk

def main():
  
  # Parse arguments
  parser = argparse.ArgumentParser(description="Weight projection serie with 2D weights of Feldkamp cone-beam reconstruction algorithm.")
  parser.add_argument("--geometry", "-g", help="XML geometry file name", required=True)
  parser.add_argument("--path", "-p", help="Path containing projections", required=True)
  parser.add_argument("--regexp", "-r", help="Regular expression to select projection files in path", required=True)
  parser.add_argument("--output", "-o", help="Output file name", required=True)
  parser.add_argument("--divisions", "-d", help="Number of stream divisions to cope with large serie", required=False, default=1, type=int)
  parser.add_argument("--verbose", "-v", help="Verbose execution", type=bool)
  args=parser.parse_args()
  
  # Generate file names
  names = itk.RegularExpressionSeriesFileNames.New()
  names.SetDirectory(args.path)
  names.SetNumericSort(False)
  names.SetRegularExpression(args.regexp)
  names.SetSubMatch(0)

  if args.verbose: print(f"Regular expression matches {len(names.GetFileNames())} file(s)...")

  # Projections reader
  ProjectionImageType = itk.Image[itk.F, 4]
  ReaderType = rtk.ProjectionsReader[ProjectionImageType]
  reader = ReaderType.New()
  reader.SetFileNames(names.GetFileNames())
  reader.GenerateOutputInformation()

  # Geometry
  if args.verbose: print(f"Reading geometry information from {args.geometry}...")
  geometryReader = rtk.ThreeDCircularProjectionGeometryXMLFileReader.New()
  geometryReader.SetFilename(args.geometry)
  geometryReader.GenerateOutputInformation()
  
  # Weight filter
  WeightType = pct.FDKDDWeightProjectionFilter[ProjectionImageType]
  wf = WeightType.New()
  wf.SetInput(reader.GetOutput())
  wf.SetGeometry(geometryReader.GetOutputObject())
  wf.InPlaceOff()

  # Writer
  WriterType = itk.ImageFileWriter[ProjectionImageType]
  writer = WriterType.New()
  writer.SetFileName(args.output)
  writer.SetInput(wf.GetOutput())
  writer.SetNumberOfStreamDivisions(args.divisions)
  writer.Update()


if __name__ == '__main__':
  main()